src/txmempool.cpp

```diff
@@ -837,3 +845,110 @@ const CTxMemPool::setEntries & CTxMemPool::GetMemPoolChildren(txiter entry) cons
     assert(it != mapLinks.end());
     return it->second.children;
 }
+
+CFeeRate CTxMemPool::GetMinFee() const {
+    LOCK(cs);
+    if (!blockSinceLastRollingFeeBump)
+        return CFeeRate(rollingMinimumFeeRate);
+
+    int64_t time = GetTime();
+    if (time > lastRollingFeeUpdate + 10) {
+        rollingMinimumFeeRate = rollingMinimumFeeRate / pow(2.0, (time - lastRollingFeeUpdate) / ROLLING_FEE_HALFLIFE);
+        lastRollingFeeUpdate = time;
+    }
+    return CFeeRate(rollingMinimumFeeRate);
+}
+
+void CTxMemPool::trackRemovedOrAddFailed(const CFeeRate& rate) {
+    AssertLockHeld(cs);
+    if (rate.GetFeePerK() > rollingMinimumFeeRate) {
+        rollingMinimumFeeRate = rate.GetFeePerK();
+        blockSinceLastRollingFeeBump = false;
+    }
+}
+
+bool CTxMemPool::TrimToSize(size_t sizelimit, const CTxMemPoolEntry& toadd) {
+    LOCK(cs);
+
+    CFeeRate bestFeeRateRemoved;
+    setEntries stage;
+    std::set<uint256> protect;
+    BOOST_FOREACH(const CTxIn& in, toadd.GetTx().vin)
+        protect.insert(in.prevout.hash);
```

Comment:
If you change ```TrimToSize()``` to take as an argument the ancestors of the entry being considered (which is calculated earlier in ```AcceptToMemoryPool()```, then you can get rid of ```protect```, and instead just check that each package root that you consider isn't an ancestor of the entry being added.  (This is what I did in #6557 and I think it helps make the code a lot simpler, especially combined with using ```CalculateDescendants()``` to grab all the descendants instead of writing a new loop here.)